reference/src/layout/unions.md

```diff
@@ -6,29 +6,109 @@ not to change until an RFC ratifies them.
 
 [#13]: https://github.com/rust-rfcs/unsafe-code-guidelines/issues/13
 
-The only degree of freedom the compiler has when computing the layout of a union
-like
+### Layout of individual union fields
+
+A union consists of several variants, one for each field. All variants have the
+same size and start at the same memory address, such that in memory the variants
+overlap. This can be visualized as follows:
 
 ```rust,ignore
-union U { f1: T1, f2: T2 }
+[ <--> [field0_ty] <----> ]
+[ <----> [field1_ty] <--> ]
+[ <---> [field2_ty] <---> ]
 ```
 
-is to determine the offset of the fields. The layout of these fields themselves
-is already entirely determined by their types, and since we intend to allow
-creating references to fields (`&u.f1`), unions do not have any wiggle-room
-there.
+> **Figure: union field layout**: Each row in the picture shows the layout of
+> the union for each of its fields, where the square brackets `[]` depict an
+> array of bytes, `<-...->` denotes different amount of padding, and
+> `[field{i}_ty]` is the bytes of the type of the `i`-th union field.
+
+The individual fields (`[field{i}_ty_]`) are blocks of fixed size determined by
+the field's layout. The only degrees of freedom the compiler has when computing
+the layout of a union are the size of the union, which can be larger than the
+size of its largest field, and the offset of each union field within its
+variant. How these are picked depends on certain constraints, lik for example,
+the alignment requirements of the fields, the `#[repr]` attribute of the
+`union`, etc.
+
+### Unions with default layout ("`repr(Rust)`")
+
+Except for the guarantees provided below for some specific cases, the default
+layout of Rust unions is, _in general_, **unspecified**.
 
-### Default layout ("repr rust")
+That is, there are no _general_ guarantees about the offset of the fields,
+whether all fields have the same offset, etc.
 
-**The default layout of unions is not specified.** As of this writing, we want
-to keep the option of using non-zero offsets open for the future; whether this
-is useful depends on what exactly the compiler-assumed invariants about union
-contents are.
+<details><summary><b>Rationale</b></summary>
+
+As of this writing, we want to keep the option of using non-zero offsets open
+for the future; whether this is useful depends on what exactly the
+compiler-assumed invariants about union contents are. This might become clearer
+after the validity of unions
+([unsafe-code-guidelines/73](https://github.com/rust-lang/unsafe-code-guidelines/issues/73))
+is settled.
 
 Even if the offsets happen to be all 0, there might still be differences in the
 function call ABI.  If you need to pass unions by-value across an FFI boundary,
 you have to use `#[repr(C)]`.
 
+</details>
+
+#### Layout of unions with a single non-zero-sized field
+
+The layout of unions with a single non-zero-sized field is the same as the
```

Comment:
What about saying "single non-1-ZST-field"? Then the first condition becomes redundant and this rule even becomes more general.